hls-plugin-api/src/Ide/Plugin/RangeMap.hs

```diff
@@ -0,0 +1,44 @@
+{-# LANGUAGE DerivingStrategies         #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+-- | A map that allows fast \"in-range\" filtering. 'RangeMap' is meant
+-- to be constructed once and cached as part of a Shake rule. If
+-- not, the map will be rebuilt upon each invocation, yielding slower
+-- results compared to the list-based approach!
+module Ide.Plugin.RangeMap
+  ( RangeMap(..),
+    fromList,
+    fromList',
+    filterByRange,
+  ) where
+
+import           Data.Bifunctor                           (first)
+import           Data.Foldable                            (foldl')
+import           Development.IDE.Graph.Classes            (NFData)
+import qualified HaskellWorks.Data.IntervalMap.FingerTree as IM
+import           Language.LSP.Types                       (Position,
+                                                           Range (Range))
+
+-- | A map from code ranges to values.
+newtype RangeMap a = RangeMap
+  { unRangeMap :: IM.IntervalMap Position a
+    -- ^ 'IM.Interval' of 'Position' corresponds to a 'Range'
+  }
+  deriving newtype (NFData)
+
+-- | Construct a 'RangeMap' from a 'Range' accessor and a list of values.
+fromList :: (a -> Range) -> [a] -> RangeMap a
+fromList extractRange = fromList' . map (\x -> (extractRange x, x))
+
+fromList' :: [(Range, a)] -> RangeMap a
+fromList' = RangeMap . toIntervalMap . map (first rangeToInterval)
+  where
+    toIntervalMap :: Ord v => [(IM.Interval v, a)] -> IM.IntervalMap v a
+    toIntervalMap = foldl' (\m (i, v) -> IM.insert i v m) IM.empty
+
+-- | Filter a 'RangeMap' by a given 'Range'.
+filterByRange :: Range -> RangeMap a -> [a]
+filterByRange range = map snd . IM.dominators (rangeToInterval range) . unRangeMap
+
+rangeToInterval :: Range -> IM.Interval Position
+rangeToInterval (Range s e) = IM.Interval s e
```

Comment:
This is, strictly, wrong! LSP `Range`s have _exclusive_ upper bounds, whereas the intervals here are supposed to be closed (i.e. inclusive at both ends). I think converting a `Range` to a closed interval may be difficult if you need to drop down a line (i.e. if the end of the `Range` is `(x,0)`), because you'd need to know where the end of the previous line is. Not sure what to do there or whether this is a problem but it definitely needs a comment.

Reply:
I couldn't say something about it right away because the test data I generated at the time didn't have any multiline ranges. However I extended the generators now, and the tests still pass (that is, `RangeMap` behaves identical to the list-based approach with multiline ranges as well). I think the reason for this is, even if `rangeToInterval` isn't a correct 1:1 conversion by itself, it is used for both the construction of the `RangeMap` and during the actual filtering (`filterByRange`), so it still behaves identical to the list approach. You can say it is a "two wrongs make a right" situation :smile:

Reply:
I think regardless this needs a comment! Future people will thank us! Even if it's just "this is wrong but it empirically works and I'm not sure why"